LamestWebserver/LamestWebserver/Caching/ResponseCache.cs

```diff
@@ -0,0 +1,459 @@
+                {
+                    CurrentStringResponseCacheSize -= (ulong)result.Response.Length;
+                    StringResponses.Remove(key);
+                }
+
+                response = null;
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Sets a response to the cache.
+        /// </summary>
+        /// <param name="key">The key of the response.</param>
+        /// <param name="response">The response.</param>
+        /// <param name="refreshTime">The lifetime of this entry. If null this entry doesn't have to be refreshed.</param>
+        public void SetCachedStringResponse(string key, string response, TimeSpan? refreshTime = null)
+        {
+            if (key == null)
+                throw new ArgumentNullException(nameof(key));
+
+            if (response == null)
+                throw new ArgumentNullException(nameof(response));
+
+            if (MaximumStringResponseCacheSize.HasValue && (ulong)response.Length > MaximumStringResponseCacheSize.Value)
+            {
+                Logger.LogDebugExcept($"The given response for '{key}' is to big to be cached.");
+                return;
+            }
+
+            if (MaximumStringResponseCacheSize.HasValue && CurrentStringResponseCacheSize + (ulong)response.Length > MaximumStringResponseCacheSize.Value)
+                MakeRoom((ulong)response.Length);
+
+            ResponseCacheEntry<string> entry = new ResponseCacheEntry<string>()
+            {
+                LastUpdatedDateTime = DateTime.UtcNow,
+                RefreshTime = refreshTime,
+                Response = response
+            };
+
+            using (StringResponseLock.LockWrite())
+            {
+                CurrentStringResponseCacheSize += (ulong)entry.Response.Length;
+
+                StringResponses[key] = entry;
+            }
+        }
+
+        /// <summary>
+        /// Retrueves a cached string or caches it using the source function.
+        /// </summary>
+        /// <param name="key">The key of the response.</param>
+        /// <param name="sourceIfNotCached">A function that retrieves the source if it's not cached yet.</param>
+        /// <param name="refreshTime">The lifetime of this entry. If null this entry doesn't have to be refreshed. This lifetime will not override the original lifetime of this entry.</param>
+        /// <returns>The cached response.</returns>
+        public virtual string GetCachedString(string key, Func<string> sourceIfNotCached, TimeSpan? refreshTime = null)
+        {
+            if (key == null)
+                throw new ArgumentNullException(nameof(key));
+
+            if (GetCachedStringResponse(key, out string response))
+            {
+                return response;
+            }
+            else
+            {
+                if (sourceIfNotCached == null)
+                    throw new ArgumentNullException(nameof(sourceIfNotCached));
```

Comment:
i am not sure how happy i am about the check happening here - should probably be moved to the top, or null could be returned.